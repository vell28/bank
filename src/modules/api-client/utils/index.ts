import {
  propSatisfies,
  ifElse,
  is,
  and,
  lte,
  gte,
  always,
  propEq,
  equals,
  either,
  pipe,
  head,
  propOr,
  pathOr,
  pathEq,
} from 'ramda';
import { stringify } from 'qs';

import {
  IError, IFetchResponse, IOptions, IRequestParams, ResponseAPI
} from '../entities';
import { appErrorStatuses, METHODS } from './const';

const inInterval = (start: number, stop: number) =>
  ifElse(is(Number), (status: number) => and(gte(status, start), lte(status, stop)), always(false));

export const is20x = propSatisfies(inInterval(200, 299), 'status');
export const is40x = propSatisfies(inInterval(400, 499), 'status');
export const is50x = propSatisfies(inInterval(500, 599), 'status');

export const isNetworkError = propEq('status', -1);
export const isGET = equals(METHODS.GET);

export const getConfirmFromResponseToken = (error: IError): string =>
  pipe(propOr([], 'errors'), head, pathOr('', ['properties', 'confirmationToken']))(error);

export const getConfirmCodeLengthToken = (error: IError): number =>
  pipe(propOr([], 'errors'), head, pathOr(0, ['properties', 'confirmationCodeLength']))(error);

export const getHTTPStatusCode = (error: IError): number =>
  pipe(
    propOr([], 'errors'),
    head,
    // propOr(0, 'code')
    // red error unknown not assignable to number
    // even seem to be right overload propOr<number>
    pathOr(0, ['code']),
  )(error);

export const isAllowBodyMethod = either(equals(METHODS.POST), equals(METHODS.PATCH));

export const onUnknow = (response: IFetchResponse): Promise<IError> =>
  // eslint-disable-next-line prefer-promise-reject-errors
  Promise.reject({
    ok: false,
    status: appErrorStatuses.UNKNOWN.code,
    errors: [
      {
        code: appErrorStatuses.UNKNOWN.code,
        type: response.statusText,
        message: 'Unknown fetch error.',
      },
    ],
  });

export const makeUrl = (method: string, url: string, data?: any): string =>
  isGET(method) && data ? `${url}${stringify(data, { addQueryPrefix: true })}` : url;

export const makeRequestOptions = (
  method: string = METHODS.GET,
  { data, headers, isFormData }: IRequestParams,
): IOptions => {
  const options = {
    method,
    headers,
  };

  if (isAllowBodyMethod(method) && data) {
    let body: any;
    if (isFormData) {
      body = new FormData();
      // eslint-disable-next-line no-restricted-syntax
      for (const k in data) {
        // eslint-disable-next-line no-prototype-builtins
        if (data.hasOwnProperty(k)) {
          body.append(k, data[k]);
        }
      }
      body = new URLSearchParams(body);
    } else {
      body = JSON.stringify(data);
    }
    return {
      ...options,
      body,
    };
  }

  return options;
};

export const isSuccessResponse = (resp: ResponseAPI<any>): boolean => pathEq(['data', 'status'], 'ok')(resp);
