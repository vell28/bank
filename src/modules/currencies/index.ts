export enum CurrencyCodes {
  USD = 'USD',
  EUR = 'EUR',
  GBP = 'GBP',
  JPY = 'JPY',
  RUB = 'RUB',
  CHF = 'CHF',
  BTC = 'BTC',
  NZD = 'NZD',
  DKK = 'DKK',
  SGD = 'SGD',
  HKD = 'HKD',
  PLN = 'PLN',
  AUD = 'AUD',
  SEK = 'SEK',
  NOK = 'NOK',
  CZK = 'CZK',
  MXN = 'MXN',
  CNH = 'CNH',
}

export type CurrencyCodeType =
  | CurrencyCodes.EUR
  | CurrencyCodes.GBP
  | CurrencyCodes.JPY
  | CurrencyCodes.RUB
  | CurrencyCodes.USD
  | CurrencyCodes.CHF
  | CurrencyCodes.BTC
  | CurrencyCodes.NZD
  | CurrencyCodes.DKK
  | CurrencyCodes.SGD
  | CurrencyCodes.HKD
  | CurrencyCodes.PLN
  | CurrencyCodes.AUD
  | CurrencyCodes.SEK
  | CurrencyCodes.NOK
  | CurrencyCodes.CZK
  | CurrencyCodes.MXN
  | CurrencyCodes.CNH;

export enum CurrencyNames {
  USD = 'Dollar',
  EUR = 'Euro',
  GBP = 'Pound Sterling',
  JPY = 'Yen',
  RUB = 'Ruble',
  CHF = 'Swiss franc',
  BTC = 'Bitcoin',
  NZD = 'NZD',
  DKK = 'DKK',
  SGD = 'SGD',
  HKD = 'HKD',
  PLN = 'PLN',
  AUD = 'AUD',
  SEK = 'SEK',
  NOK = 'NOK',
  CZK = 'CZK',
  MXN = 'MXN',
  CNH = 'CNH',
}

export type CurrencyNameType =
  | CurrencyNames.USD
  | CurrencyNames.EUR
  | CurrencyNames.GBP
  | CurrencyNames.JPY
  | CurrencyNames.RUB
  | CurrencyNames.CHF
  | CurrencyNames.BTC
  | CurrencyNames.NZD
  | CurrencyNames.DKK
  | CurrencyNames.SGD
  | CurrencyNames.HKD
  | CurrencyNames.PLN
  | CurrencyNames.AUD
  | CurrencyNames.SEK
  | CurrencyNames.NOK
  | CurrencyNames.CZK
  | CurrencyNames.MXN
  | CurrencyNames.CNH;

export enum CurrencySymbols {
  USD = '$',
  EUR = '€',
  GBP = '£',
  JPY = '¥',
  RUB = '₽',
  CHF = 'CHF',
  BTC = 'BTC',
  NZD = 'NZD',
  DKK = 'DKK',
  SGD = 'SGD',
  HKD = 'HKD',
  PLN = 'PLN',
  AUD = 'AUD',
  SEK = 'SEK',
  NOK = 'NOK',
  CZK = 'CZK',
  MXN = 'MXN',
  CNH = 'CNH',
}

export type CurrencySymbolType =
  | CurrencySymbols.USD
  | CurrencySymbols.EUR
  | CurrencySymbols.GBP
  | CurrencySymbols.JPY
  | CurrencySymbols.RUB
  | CurrencySymbols.CHF
  | CurrencySymbols.BTC
  | CurrencySymbols.NZD
  | CurrencySymbols.DKK
  | CurrencySymbols.SGD
  | CurrencySymbols.HKD
  | CurrencySymbols.PLN
  | CurrencySymbols.AUD
  | CurrencySymbols.SEK
  | CurrencySymbols.NOK
  | CurrencySymbols.CZK
  | CurrencySymbols.MXN
  | CurrencySymbols.CNH;

export const CURRENCIES_IN_PROJECT: ICurrency[] = [
  {
    name: CurrencyNames.USD,
    symbol: CurrencySymbols.USD,
    code: CurrencyCodes.USD,
  },
  {
    name: CurrencyNames.EUR,
    symbol: CurrencySymbols.EUR,
    code: CurrencyCodes.EUR,
  },
  {
    name: CurrencyNames.GBP,
    symbol: CurrencySymbols.GBP,
    code: CurrencyCodes.GBP,
  },
  {
    name: CurrencyNames.JPY,
    symbol: CurrencySymbols.JPY,
    code: CurrencyCodes.JPY,
  },
  {
    name: CurrencyNames.RUB,
    symbol: CurrencySymbols.RUB,
    code: CurrencyCodes.RUB,
  },
  {
    name: CurrencyNames.CHF,
    symbol: CurrencySymbols.CHF,
    code: CurrencyCodes.CHF,
  },
  {
    name: CurrencyNames.BTC,
    symbol: CurrencySymbols.BTC,
    code: CurrencyCodes.BTC,
  },
  {
    name: CurrencyNames.NZD,
    symbol: CurrencySymbols.NZD,
    code: CurrencyCodes.NZD,
  },
  {
    name: CurrencyNames.DKK,
    symbol: CurrencySymbols.DKK,
    code: CurrencyCodes.DKK,
  },
  {
    name: CurrencyNames.SGD,
    symbol: CurrencySymbols.SGD,
    code: CurrencyCodes.SGD,
  },
  {
    name: CurrencyNames.HKD,
    symbol: CurrencySymbols.HKD,
    code: CurrencyCodes.HKD,
  },
  {
    name: CurrencyNames.PLN,
    symbol: CurrencySymbols.PLN,
    code: CurrencyCodes.PLN,
  },
  {
    name: CurrencyNames.AUD,
    symbol: CurrencySymbols.AUD,
    code: CurrencyCodes.AUD,
  },
  {
    name: CurrencyNames.SEK,
    symbol: CurrencySymbols.SEK,
    code: CurrencyCodes.SEK,
  },
  {
    name: CurrencyNames.NOK,
    symbol: CurrencySymbols.NOK,
    code: CurrencyCodes.NOK,
  },
  {
    name: CurrencyNames.CZK,
    symbol: CurrencySymbols.CZK,
    code: CurrencyCodes.CZK,
  },
  {
    name: CurrencyNames.MXN,
    symbol: CurrencySymbols.MXN,
    code: CurrencyCodes.MXN,
  },
  {
    name: CurrencyNames.CNH,
    symbol: CurrencySymbols.CNH,
    code: CurrencyCodes.CNH,
  },
];

export interface ICurrency {
  name: CurrencyNameType;
  symbol: CurrencySymbolType;
  code: CurrencyCodeType;
}

export interface ICurrencyValue {
  type: ICurrency;
  value: string;
}
