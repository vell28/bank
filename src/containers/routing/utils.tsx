import {
  findIndex, isNil, pipe, pathOr, forEach
} from 'ramda';
import { matchPath } from 'react-router';
import { RouteComponentProps } from 'react-router-dom';
import { connectedRouterRedirect } from 'redux-auth-wrapper/history4/redirect';

import { CurrencyCodeType } from 'modules/currencies';
import { getToken, getFirebaseToken } from 'models/authorization/redux/selectors';
import { IStore } from 'modules/store/types';

enum RoutePaths {
  BANKING = 'banking',
  BANKING_ACCOUNT = 'banking_account',
  NEW_ACCOUNT = 'new_account',
  BONUSES = 'bonuses',
  BUSINESS = 'business',
  INSURANCE = 'insurance',
  CRYPTOTRADING = 'cryptotrading',
  PROFILE = 'profile',
  AUTHORIZATION = 'authorization',
  SIGNIN = 'signin',
  SIGNIN_CONFIRM = 'signin_confirm',
  TOPUP_PUBLIC = 'topup_public',
}

export type RoutePathType =
  | RoutePaths.BANKING
  | RoutePaths.BANKING_ACCOUNT
  | RoutePaths.NEW_ACCOUNT
  | RoutePaths.BONUSES
  | RoutePaths.BUSINESS
  | RoutePaths.CRYPTOTRADING
  | RoutePaths.PROFILE
  | RoutePaths.AUTHORIZATION
  | RoutePaths.SIGNIN
  | RoutePaths.TOPUP_PUBLIC
  | RoutePaths.INSURANCE
  | RoutePaths.SIGNIN_CONFIRM;

export const BANKING_PATH = '/';
export const BANKING_ACCOUNT_PATH = '/banking/account/:id/:currency';
export const NEW_ACCOUNT_PATH = '/create-account';
export const BONUSES_PATH = '/bonuses';
export const BUSINESS_PATH = '/business';
export const INSURANCE_PATH = '/insurance';
export const CRYPTOTRADING_PATH = 'https://frontend.stage.ak-ecosystem.com/';
export const PROFILE_PATH = '/profile';
export const AUTHORIZATION_PATH = '/authorization';
export const SIGNIN_PATH = '/authorization/signin';
export const SIGNIN_CONFIRM_PATH = '/authorization/signin/confirm';
export const ERROR_PATH = '/error';
export const TOPUP_PUBLIC_PATH = '/public-topup/:id';

export interface IPath {
  path: string;
  name: RoutePathType;
}

export interface IPathConfig {
  [key: string]: IPath;
}

export const PATHS_CONFIG: IPathConfig = {
  [RoutePaths.SIGNIN]: {
    path: SIGNIN_PATH,
    name: RoutePaths.SIGNIN,
  },
  [RoutePaths.SIGNIN_CONFIRM]: {
    path: SIGNIN_CONFIRM_PATH,
    name: RoutePaths.SIGNIN_CONFIRM,
  },
  [RoutePaths.AUTHORIZATION]: {
    path: AUTHORIZATION_PATH,
    name: RoutePaths.AUTHORIZATION,
  },
  [RoutePaths.BANKING]: {
    path: BANKING_PATH,
    name: RoutePaths.BANKING,
  },
  [RoutePaths.BANKING_ACCOUNT]: {
    path: BANKING_ACCOUNT_PATH,
    name: RoutePaths.BANKING_ACCOUNT,
  },
  [RoutePaths.NEW_ACCOUNT]: {
    path: NEW_ACCOUNT_PATH,
    name: RoutePaths.NEW_ACCOUNT,
  },
  [RoutePaths.BONUSES]: {
    path: BONUSES_PATH,
    name: RoutePaths.BONUSES,
  },
  [RoutePaths.BUSINESS]: {
    path: BUSINESS_PATH,
    name: RoutePaths.BUSINESS,
  },
  [RoutePaths.INSURANCE]: {
    path: INSURANCE_PATH,
    name: RoutePaths.INSURANCE,
  },
  [RoutePaths.CRYPTOTRADING]: {
    path: CRYPTOTRADING_PATH,
    name: RoutePaths.CRYPTOTRADING,
  },
  [RoutePaths.PROFILE]: {
    path: PROFILE_PATH,
    name: RoutePaths.PROFILE,
  },
  [RoutePaths.TOPUP_PUBLIC]: {
    path: TOPUP_PUBLIC_PATH,
    name: RoutePaths.TOPUP_PUBLIC,
  },
};

export const mainMenuPaths: IPath[] = [
  PATHS_CONFIG.banking,
  PATHS_CONFIG.bonuses,
  PATHS_CONFIG.business,
  PATHS_CONFIG.cryptotrading,
  PATHS_CONFIG.insurance,
  PATHS_CONFIG.profile,
];

const pathWithParamsForMatch: string[] = [BANKING_ACCOUNT_PATH];

export const getActiveIndex = (location: string, paths: IPath[]) =>
  pipe(
    findIndex((item: IPath) => {
      const match = matchPath(location, {
        path: item.path,
        exact: true,
      });
      return !isNil(match);
    }),
    (index: number) => (index < 0 ? 0 : index),
  )(paths);

export const getBankingAccountPath = (id: string, currency: CurrencyCodeType): string =>
  `/banking/account/${id}/${currency}`;

export const getPathParamValue = (key: string, match: string, pathList: string[] = pathWithParamsForMatch): string => {
  let matchedPath = {};
  forEach((path: string) => {
    const res = matchPath(match, {
      path,
      exact: true,
      strict: false,
    });
    if (res !== null) {
      matchedPath = res;
    }
  }, pathList);

  return pathOr('', ['params', key])(matchedPath);
};

export const getRouteParamByName = (name: string, routeProps: RouteComponentProps): string =>
  pathOr('', ['match', 'params', name], routeProps);

export const userIsAuthenticated = connectedRouterRedirect({
  redirectPath: AUTHORIZATION_PATH,
  authenticatedSelector: (store: IStore) => !!getFirebaseToken(store) || !!getToken(store),
  wrapperDisplayName: 'UserIsAuthenticated',
});
