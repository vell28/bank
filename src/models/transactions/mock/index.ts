import { CurrencyCodes } from 'modules/currencies';
import { ITransaction, ITransactionResponse } from '../entities';

export const transactionData: ITransaction = {
  id: 'PPY184598',
  amount: -10.0,
  fee: 0,
  status: 'COMPLETED',
  paymentType: 'TRANSFER',
  fromNumber: 'MT98PAPY36836000002676370004519',
  fromName: 'Ralf Williams',
  fromBank: 'Papaya',
  toNumber: 'MT97PAPY36836000002676370003470',
  toName: 'Dmitrii Popov',
  toBank: 'Papaya',
  currency: {
    code: CurrencyCodes.EUR,
  },
  clientId: '6449',
  accountId: 'PPY5107',
  description: 'To: Dmitrii Popov Vvg',
  executedAt: '2019-10-11T10:12:00',
  referenceNumber: 'PPY1e1d31e3-6acb-4d60-9f70-b9151221d11',
  paymentToRepeat: {
    type: 'PAYMENT_PHONE',
    fromPhone: '79111111111',
    account: 'PPY5107',
    fromCardPan: null,
    fromCardId: null,
    beneficiaryName: null,
    beneficiaryAccount: null,
    country: null,
    city: null,
    address: null,
    swiftCode: null,
    beneficiaryBank: null,
    intermediaryBank: null,
    intermediarySwift: null,
    intermediaryAccount: null,
    commissionType: null,
    urgency: null,
    iban: null,
    phoneNumber: '79085491121',
    cardNumber: null,
    currencyCode: CurrencyCodes.EUR,
    amount: 10,
    purpose: 'other Vvgh',
    transferDetails: null,
  },
  openapiData: {
    type: 'TO_CONTACT',
    toCard: null,
    toPhone: '79085491121',
    subType: null,
    externalOwnerID: null,
    fromCardPan: null,
  },
  operationType: null,
  fromPanDisplay: null,
};

const second: ITransaction = {
  id: 'PPY184597',
  amount: -1.0,
  fee: 0,
  status: 'COMPLETED',
  paymentType: 'TRANSFER',
  fromNumber: 'MT98PAPY36836000002676370004519',
  fromName: 'Ralf Williams',
  fromBank: 'Papaya',
  toNumber: 'MT52PAPY36836000002676370000303',
  toName: 'Elfur Long',
  toBank: 'Papaya',
  currency: {
    code: CurrencyCodes.EUR,
  },
  clientId: '6449',
  accountId: 'PPY5107',
  description: 'To: Elfur Long 12',
  executedAt: '2019-10-11T10:08:47',
  referenceNumber: 'PPY84f7087c-c511-4928-b299-13ee47c1bd9',
  paymentToRepeat: {
    type: 'PAYMENT_PHONE',
    fromPhone: '79111111111',
    account: 'PPY5107',
    fromCardPan: null,
    fromCardId: null,
    beneficiaryName: null,
    beneficiaryAccount: null,
    country: null,
    city: null,
    address: null,
    swiftCode: null,
    beneficiaryBank: null,
    intermediaryBank: null,
    intermediarySwift: null,
    intermediaryAccount: null,
    commissionType: null,
    urgency: null,
    iban: null,
    phoneNumber: '79032950311',
    cardNumber: null,
    currencyCode: CurrencyCodes.EUR,
    amount: 1,
    purpose: 'other 123',
    transferDetails: null,
  },
  openapiData: {
    type: 'TO_CONTACT',
    toCard: null,
    toPhone: '79032950311',
    subType: null,
    externalOwnerID: null,
    fromCardPan: null,
  },
  operationType: null,
  fromPanDisplay: null,
};

export const transactionList = [transactionData, second];

export const transactionResponse: ITransactionResponse = {
  content: transactionList,
  maxPageNumber: 10,
  pageSize: 20,
  pageNumber: 1,
  length: 0,
};
