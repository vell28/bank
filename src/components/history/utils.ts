import { isToday, isYesterday, format } from 'date-fns';
import { pathOr } from 'ramda';

import { ITransactionList, ITransaction, IOrderedTransactions } from 'models/transactions/entities';

export const getDayTitle = (date: Date): string => {
  if (isToday(date)) {
    return 'today';
  }
  if (isYesterday(date)) {
    return 'yesterday';
  }
  return format(date, 'dd/MM/yyyy');
};

export const groupByAccountType = (list: ITransactionList): IOrderedTransactions => {
  const ordered: IOrderedTransactions = {};
  list.map((item: ITransaction) => {
    const type = pathOr('unknown', ['simpleAccount', 'accountType'])(item);
    if (ordered[type]) {
      ordered[type].push(item);
    } else {
      ordered[type] = [item];
    }
    return item;
  });
  return ordered;
};
