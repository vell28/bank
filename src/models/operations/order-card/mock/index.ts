import { IServerAddress } from 'models/organizations/redux/actions/entities';
import { IClient } from 'models/organizations/entities';
import { IAddress, ICardOrderOwner } from '../entities';

export const exampleAddress: IAddress = {
  countryCode: 'DE',
  city: 'Cologne',
  postalCode: '50667',
  street: 'Domkloster',
  streetNumber: '4',
  apartmentNumber: 0,
};

export const exampleOrganizationAddress: IAddress = {
  countryCode: 'DE',
  city: 'Cologne',
  postalCode: '50667',
  street: 'Domkloster',
  streetNumber: '4',
  apartmentNumber: 0,
};

export const exampleClientInfo: ICardOrderOwner = {
  embossedName: 'Frank-Walter Steinmeier',
  phoneNumber: '9111111',
};

export const exampleServerAddress: IServerAddress = {
  country: 'Germany',
  province: 'North Rhine-Westphalia',
  city: 'Düsseldorf',
  postalCode: '40479',
  street: 'Gartenstraße',
  streetNumber: '53',
  apartmentNumber: '42',
};

export const exampleClient: IClient = {
  address: exampleOrganizationAddress,
  name: 'Frank-Walter Steinmeier',
  phoneNumber: '9111111',
};
