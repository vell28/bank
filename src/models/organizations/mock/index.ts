import { CurrencyCodes } from 'modules/currencies';
import {
  IAccount, AccountStatus, AccountTypes, AccountColors
} from '../entities';
import { ICard, CardProduct, Limit } from '../../operations/card-settings/entities';

import { IServerClient } from '../redux/actions/entities';
import { exampleServerAddress } from '../../operations/order-card/mock';

export const cardData: ICard = {
  accountId: 'PPY5107',
  availableBalance: 1897.61,
  balance: 0,
  balanceHold: 0,
  balances: [
    {
      currency: CurrencyCodes.EUR,
      balance: 1030.05,
      balanceHold: 0,
      availableBalance: 1897.61,
    },
  ],
  clientId: '6449',
  currency: { code: CurrencyCodes.EUR },
  expireAt: '2022-05-01T00:00:00',
  id: 'PPY13643',
  limits: [
    {
      type: Limit.DAY,
      title: null,
      maxValue: 500,
      usedValue: 0,
    },
  ],
  number: '5277 60_3 984',
  options: {
    limits: { disable: false },
  },
  owner: { embossedName: 'RALF WILLIAMS' },
  paySystem: 'MASTERCARD',
  plasticStatus: 'PENDING_SHIPMENT',
  productType: CardProduct.MAIN,
  status: AccountStatus.ACTIVE,
};

export const accountsData: IAccount[] = [
  {
    accountName: 'MT98PAPY36836000002676370004519',
    accountType: AccountTypes.PHYSICAL,
    accruedInterest: null,
    availableBalance: 1897.61,
    availableOperations: [],
    balance: 0,
    balanceHold: 0,
    color: AccountColors.DEFAULT,
    balances: [
      {
        currency: CurrencyCodes.EUR,
        balance: 1030.05,
        balanceHold: 0,
        availableBalance: 1897.61,
      },
    ],
    bonusProgramActivate: false,
    clientId: '6449',
    createdAt: '2019-11-24T14:59:55',
    currency: { code: CurrencyCodes.EUR },
    id: 'PPY5107',
    number: 'MT98PAPY36836000002676370004519',
    status: AccountStatus.ACTIVE,
  },
  {
    accountName: 'MT98PAPY36836000002676370004519',
    accountType: AccountTypes.JURIDICAL,
    accruedInterest: null,
    availableBalance: 1897.61,
    availableOperations: [],
    balance: 0,
    balanceHold: 0,
    color: AccountColors.DEFAULT,
    balances: [
      {
        currency: CurrencyCodes.EUR,
        balance: 1030.05,
        balanceHold: 0,
        availableBalance: 1897.61,
      },
    ],
    bonusProgramActivate: false,
    clientId: '6449',
    createdAt: '2019-11-24T14:59:55',
    currency: { code: CurrencyCodes.EUR },
    id: 'PPY6666',
    number: 'MT98PAPY36836000002676370004519',
    status: AccountStatus.ACTIVE,
  },
];

export const clientData: IServerClient = {
  id: '42',
  name: 'Test Client',
  citizenship: 'Germany',
  email: 'test.client@test.com',
  phoneNumber: '222333',
  address: exampleServerAddress,
};
