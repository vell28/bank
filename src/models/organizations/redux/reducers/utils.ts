import {
  findIndex, pathOr, pipe, propEq
} from 'ramda';

import { IAddress } from 'models/operations/order-card/entities';
import { IOrganization } from 'models/organizations/entities';
import getDataServer from '../../../../modules/countries/country-list.json';
import { IServerAddress, IServerOrganization } from '../actions/entities.js';

export const findAccountIndexById = (id: string) =>
  pipe(pathOr([], ['data', 0, 'accounts']), findIndex(propEq('id', id)));

// alias
type countryCode = string;

const parseCountry = (country: string): countryCode => {
  const targetCountry = getDataServer.find(propEq('name', country));
  if (targetCountry) {
    return targetCountry.code;
  }
  throw new Error(`Unknown country ${country}`); // nothing else can do
};

const parseApartmentNumber = (serverRepresentation: string): number => {
  const parsed = parseInt(serverRepresentation, 10);
  // eslint-disable-next-line no-restricted-globals
  if (!isNaN(parsed)) {
    return parsed;
  }
  return 0;
  // throw new Error('Invalid apartmentNumber');
};

// inthis context client means client-side(front-end) application
const parseSingleAddressToClientType = ({
  country = '',
  province,
  city = '',
  postalCode = '',
  street = '',
  streetNumber = '',
  apartmentNumber = '0',
}: IServerAddress): IAddress => ({
  countryCode: parseCountry(country),
  province,
  city,
  postalCode,
  street,
  streetNumber,
  apartmentNumber: parseApartmentNumber(apartmentNumber),
});

export const parseOrganizationsToClientType = (organizations: IServerOrganization[]): IOrganization[] => {
  return organizations.map((item) => {
    return {
      ...item,
      client: {
        ...item.client,
        // we expect server to send us an address
        address: parseSingleAddressToClientType(item.client.address as IServerAddress),
      },
    };
  });
};
