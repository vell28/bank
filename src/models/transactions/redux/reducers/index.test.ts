import { isNotEmpty } from 'utils/ramda';
import { CurrencyCodes } from 'modules/currencies';
import { transactions, ITransactionsState } from '.';
import {
  TOGGLE_ACTION_BTNS,
  SET_TRANSACTION,
  TOGGLE_HISTORY_FILTER,
  TOGGLE_HISTORY_FILTER_BY_TYPE,
  TOGGLE_HISTORY_FILTER_BY_DATE,
  TOGGLE_RECEIPT,
  ISetTransactionsAction,
  IToggleBtnsAction,
  IReceiptAction,
  ISetHistoryFilterAction,
  IToggleHistoryFilterByDateAction,
} from '../actions';

import { transactionResponse } from '../../mock';

const initTransaction: ITransactionsState = {
  isLoading: false,
  isSuccess: false,
  isError: false,
  data: {},
  pagination: {
    maxPageNumber: 0,
    pageSize: 30,
    pageNumber: 0,
    length: 0,
  },
  filter: {
    isShown: false,
  },
  settings: {
    shownBtns: {},
  },
  errors: [],
};

const transactId = 'PPY184598';

describe('transactions reducers', () => {
  test('transactions/SET_TRANSACTION', () => {
    const action: ISetTransactionsAction = {
      type: `${SET_TRANSACTION}_SUCCESS`,
      payload: { data: transactionResponse },
    };

    const newState = transactions(initTransaction, action);
    expect(isNotEmpty(newState.data)).toBeTruthy();
    expect(newState.pagination.length).toBe(transactionResponse.content.length);
    expect(newState.pagination.pageNumber).toBe(transactionResponse.pageNumber);
    expect(newState.pagination.maxPageNumber).toBe(transactionResponse.maxPageNumber);
    expect(newState.pagination.pageSize).toBe(transactionResponse.pageSize);
    expect(newState).toMatchSnapshot();
  });
  test('transactions/TOGGLE_ACTION_BTNS', () => {
    const action: IToggleBtnsAction = {
      type: TOGGLE_ACTION_BTNS,
      payload: transactId,
    };
    const newState = transactions(initTransaction, action);
    expect(newState.settings.shownBtns[transactId]).toBeTruthy();
    const nextState = transactions(newState, action);
    expect(nextState.settings.shownBtns[transactId]).toBeFalsy();
  });
  test('transactions/TOGGLE_RECEIPT', () => {
    const action: IReceiptAction = {
      type: TOGGLE_RECEIPT,
      payload: transactId,
    };
    const newState = transactions(initTransaction, action);
    expect(newState.settings.receiptId).toBe(transactId);
    const nextState = transactions(newState, action);
    expect(nextState.settings.receiptId).toBeUndefined();
  });
  test('transactions/TOGGLE_HISTORY_FILTER', () => {
    const action = {
      type: TOGGLE_HISTORY_FILTER,
    };
    const newState = transactions(initTransaction, action);
    expect(newState.filter.isShown).toBeTruthy();
    const nextState = transactions(newState, action);
    expect(nextState.filter.isShown).toBeFalsy();
  });

  test('transactions/TOGGLE_HISTORY_FILTER_BY_TYPE', () => {
    const action: ISetHistoryFilterAction = {
      type: TOGGLE_HISTORY_FILTER_BY_TYPE,
      payload: {
        currency: CurrencyCodes.EUR,
        accountType: 'PHYSICAL',
      },
    };
    const newState = transactions(initTransaction, action);
    expect(newState.filter.currency).toBe(CurrencyCodes.EUR);
    expect(newState.filter.accountType).toBe('PHYSICAL');
    const nextState = transactions(newState, action);
    // test toggle, should be undefined after next click to filter
    expect(nextState.filter.currency).toBeUndefined();
    expect(nextState.filter.accountType).toBeUndefined();
  });
  test('transactions/TOGGLE_HISTORY_FILTER_BY_DATE', () => {
    const action: IToggleHistoryFilterByDateAction = {
      type: TOGGLE_HISTORY_FILTER_BY_DATE,
      payload: '21.11.2019',
    };
    const newState = transactions(initTransaction, action);
    expect(newState.filter.date).toBe('21.11.2019');
    const nextState = transactions(newState, action);
    // test toggle, should be undefined after next click to filter
    expect(nextState.filter.date).toBeUndefined();
  });
});
