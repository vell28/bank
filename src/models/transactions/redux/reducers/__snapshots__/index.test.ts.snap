// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transactions reducers transactions/SET_TRANSACTION 1`] = `
Object {
  "data": Object {
    "11/10/2019": Array [
      Object {
        "accountId": "PPY5107",
        "amount": -10,
        "clientId": "6449",
        "currency": Object {
          "code": "EUR",
        },
        "description": "To: Dmitrii Popov Vvg",
        "executedAt": "2019-10-11T10:12:00",
        "fee": 0,
        "fromBank": "Papaya",
        "fromName": "Ralf Williams",
        "fromNumber": "MT98PAPY36836000002676370004519",
        "fromPanDisplay": null,
        "id": "PPY184598",
        "openapiData": Object {
          "externalOwnerID": null,
          "fromCardPan": null,
          "subType": null,
          "toCard": null,
          "toPhone": "79085491121",
          "type": "TO_CONTACT",
        },
        "operationType": null,
        "paymentToRepeat": Object {
          "account": "PPY5107",
          "address": null,
          "amount": 10,
          "beneficiaryAccount": null,
          "beneficiaryBank": null,
          "beneficiaryName": null,
          "cardNumber": null,
          "city": null,
          "commissionType": null,
          "country": null,
          "currencyCode": "EUR",
          "fromCardId": null,
          "fromCardPan": null,
          "fromPhone": "79111111111",
          "iban": null,
          "intermediaryAccount": null,
          "intermediaryBank": null,
          "intermediarySwift": null,
          "phoneNumber": "79085491121",
          "purpose": "other Vvgh",
          "swiftCode": null,
          "transferDetails": null,
          "type": "PAYMENT_PHONE",
          "urgency": null,
        },
        "paymentType": "TRANSFER",
        "referenceNumber": "PPY1e1d31e3-6acb-4d60-9f70-b9151221d11",
        "status": "COMPLETED",
        "toBank": "Papaya",
        "toName": "Dmitrii Popov",
        "toNumber": "MT97PAPY36836000002676370003470",
      },
      Object {
        "accountId": "PPY5107",
        "amount": -1,
        "clientId": "6449",
        "currency": Object {
          "code": "EUR",
        },
        "description": "To: Elfur Long 12",
        "executedAt": "2019-10-11T10:08:47",
        "fee": 0,
        "fromBank": "Papaya",
        "fromName": "Ralf Williams",
        "fromNumber": "MT98PAPY36836000002676370004519",
        "fromPanDisplay": null,
        "id": "PPY184597",
        "openapiData": Object {
          "externalOwnerID": null,
          "fromCardPan": null,
          "subType": null,
          "toCard": null,
          "toPhone": "79032950311",
          "type": "TO_CONTACT",
        },
        "operationType": null,
        "paymentToRepeat": Object {
          "account": "PPY5107",
          "address": null,
          "amount": 1,
          "beneficiaryAccount": null,
          "beneficiaryBank": null,
          "beneficiaryName": null,
          "cardNumber": null,
          "city": null,
          "commissionType": null,
          "country": null,
          "currencyCode": "EUR",
          "fromCardId": null,
          "fromCardPan": null,
          "fromPhone": "79111111111",
          "iban": null,
          "intermediaryAccount": null,
          "intermediaryBank": null,
          "intermediarySwift": null,
          "phoneNumber": "79032950311",
          "purpose": "other 123",
          "swiftCode": null,
          "transferDetails": null,
          "type": "PAYMENT_PHONE",
          "urgency": null,
        },
        "paymentType": "TRANSFER",
        "referenceNumber": "PPY84f7087c-c511-4928-b299-13ee47c1bd9",
        "status": "COMPLETED",
        "toBank": "Papaya",
        "toName": "Elfur Long",
        "toNumber": "MT52PAPY36836000002676370000303",
      },
    ],
  },
  "errors": Array [],
  "filter": Object {
    "isShown": false,
  },
  "isError": false,
  "isLoading": false,
  "isSuccess": true,
  "pagination": Object {
    "length": 2,
    "maxPageNumber": 10,
    "pageNumber": 1,
    "pageSize": 20,
  },
  "settings": Object {
    "shownBtns": Object {},
  },
}
`;
